src/components/button-group/button-group.ts

```diff
@@ -0,0 +1,288 @@
+  }
+
+  protected override firstUpdated() {
+    if (this.disabled) {
+      this.updateDisabledState();
+    }
+
+    // Button level selection takes priority
+    if (this._selectedItems.length && !this._selectedButtons.length) {
+      this.setSelection(this._selectedItems);
+    } else if (
+      this.selection !== 'multiple' &&
+      this._selectedButtons.length > 1
+    ) {
+      const index = this._selectedButtons.indexOf(
+        this._selectedButtons.at(-1) as IgcToggleButtonComponent
+      );
+
+      for (let i = 0; i < index; i++) {
+        this._selectedButtons[i].selected = false;
+      }
+    }
+
+    this.setMutationsObserver();
+  }
+
+  private handleClick = (event: MouseEvent) => {
+    const button = event
+      .composedPath()
+      .find(
+        (element) => element instanceof IgcToggleButtonComponent
+      ) as IgcToggleButtonComponent;
+
+    if (button) {
+      switch (this.selection) {
+        case 'single':
+          this.handleSingleSelection(button);
+          break;
+        case 'single-required':
+          this.handleSingleSelection(button, true);
+          break;
+        case 'multiple':
+          this.handleMultipleSelection(button);
+          break;
+      }
+    }
+  };
+
+  private handleSingleSelection(
+    button: IgcToggleButtonComponent,
+    required = false
+  ) {
+    const selectedButton = this._selectedButtons[0];
+
+    if (selectedButton) {
+      if (required && selectedButton.value === button.value) return;
+
+      this.emitDeselectEvent(selectedButton);
+    }
+
+    if (!required && selectedButton && selectedButton.value === button.value)
+      return;
+
+    this.emitSelectEvent(button);
+  }
+
+  private handleMultipleSelection(button: IgcToggleButtonComponent) {
+    button.selected
+      ? this.emitDeselectEvent(button)
+      : this.emitSelectEvent(button);
+  }
+
+  private emitSelectEvent(button: IgcToggleButtonComponent) {
+    const select = this.emitEvent('igcSelect', {
+      cancelable: true,
+      detail: button.value,
+    });
+
+    if (select) {
+      button.selected = true;
+    }
+  }
+
+  private emitDeselectEvent(button: IgcToggleButtonComponent) {
+    const deselect = this.emitEvent('igcDeselect', {
+      cancelable: true,
+      detail: button.value,
+    });
+
+    if (deselect) {
+      button.selected = false;
+    }
+  }
+
+  private setSelection(values: string[]) {
+    values = values.filter(
+      (v) => this.toggleButtons.map((b) => b.value).indexOf(v) !== -1
+    );
+
+    const buttons = this.toggleButtons.filter(
+      (b) => values.indexOf(b.value) !== -1
+    );
+
+    if (buttons.length) {
+      if (this.selection === 'multiple') {
+        buttons.forEach((b) => (b.selected = true));
+      } else {
+        buttons.find((b) => b.value === values.at(-1))!.selected = true;
+      }
+    }
+  }
+
+  private setMutationsObserver() {
+    this.mutationObserver = new MutationObserver(async (records, observer) => {
```

Comment:
```suggestion
    this.mutationObserver = new MutationObserver((records, observer) => {
```
The observer callback does not need to be async in this scenario.